#1117: Introducing the duplicate-finder maven plugin to catch duplicate resources and classes

markkolich opened the pull request:
@johanhaleby demonstrating how to fail the build on a duplicate class/resource found on the compile or runtime classpath.

CC // @ponziani @azzazzel

markkolich commented:
I would expect this to fail, as the Maven `duplicate-finder-plugin` is now catching duplicate classes and resources in the resulting artifacts:

```
[WARNING] Found duplicate (but equal) classes in [io.rest-assured:json-path:3.3.1-SNAPSHOT, io.rest-assured:rest-assured-common:3.3.1-SNAPSHOT, io.rest-assured:xml-path:3.3.1-SNAPSHOT]:
[WARNING]   io.restassured.exception.PathException
[WARNING]   io.restassured.internal.assertion.AssertParameter
[WARNING]   io.restassured.internal.assertion.Assertion
[WARNING]   io.restassured.internal.assertion.AssertionSupport
[WARNING]   io.restassured.internal.assertion.EndToEndQuoteFragmentEscaper
[WARNING]   io.restassured.internal.assertion.GetAtPathFragmentEscaper
[WARNING]   io.restassured.internal.assertion.HyphenQuoteFragmentEscaper
[WARNING]   io.restassured.internal.assertion.PathFragmentEscaper
[WARNING]   io.restassured.internal.classpath.ClassPathResolver
[WARNING]   io.restassured.internal.mapper.ObjectDeserializationContextImpl
[WARNING]   io.restassured.internal.path.ObjectConverter
[WARNING]   io.restassured.mapper.DataToDeserialize
[WARNING]   io.restassured.mapper.ObjectDeserializationContext
[WARNING]   io.restassured.mapper.TypeRef
[WARNING]   io.restassured.mapper.factory.ObjectMapperFactory
[WARNING]   io.restassured.mapper.resolver.ObjectMapperResolver
[WARNING] Found duplicate classes/resources in compile classpath.
[WARNING] Found duplicate (but equal) classes in [io.rest-assured:json-path:3.3.1-SNAPSHOT, io.rest-assured:rest-assured-common:3.3.1-SNAPSHOT, io.rest-assured:xml-path:3.3.1-SNAPSHOT]:
[WARNING]   io.restassured.exception.PathException
[WARNING]   io.restassured.internal.assertion.AssertParameter
[WARNING]   io.restassured.internal.assertion.Assertion
[WARNING]   io.restassured.internal.assertion.AssertionSupport
[WARNING]   io.restassured.internal.assertion.EndToEndQuoteFragmentEscaper
[WARNING]   io.restassured.internal.assertion.GetAtPathFragmentEscaper
[WARNING]   io.restassured.internal.assertion.HyphenQuoteFragmentEscaper
[WARNING]   io.restassured.internal.assertion.PathFragmentEscaper
[WARNING]   io.restassured.internal.classpath.ClassPathResolver
[WARNING]   io.restassured.internal.mapper.ObjectDeserializationContextImpl
[WARNING]   io.restassured.internal.path.ObjectConverter
[WARNING]   io.restassured.mapper.DataToDeserialize
[WARNING]   io.restassured.mapper.ObjectDeserializationContext
[WARNING]   io.restassured.mapper.TypeRef
[WARNING]   io.restassured.mapper.factory.ObjectMapperFactory
[WARNING]   io.restassured.mapper.resolver.ObjectMapperResolver
[WARNING] Found duplicate classes/resources in runtime classpath.
[WARNING] Found duplicate (but equal) classes in [io.rest-assured:json-path:3.3.1-SNAPSHOT, io.rest-assured:rest-assured-common:3.3.1-SNAPSHOT, io.rest-assured:xml-path:3.3.1-SNAPSHOT]:
[WARNING]   io.restassured.exception.PathException
[WARNING]   io.restassured.internal.assertion.AssertParameter
[WARNING]   io.restassured.internal.assertion.Assertion
[WARNING]   io.restassured.internal.assertion.AssertionSupport
[WARNING]   io.restassured.internal.assertion.EndToEndQuoteFragmentEscaper
[WARNING]   io.restassured.internal.assertion.GetAtPathFragmentEscaper
[WARNING]   io.restassured.internal.assertion.HyphenQuoteFragmentEscaper
[WARNING]   io.restassured.internal.assertion.PathFragmentEscaper
[WARNING]   io.restassured.internal.classpath.ClassPathResolver
[WARNING]   io.restassured.internal.mapper.ObjectDeserializationContextImpl
[WARNING]   io.restassured.internal.path.ObjectConverter
[WARNING]   io.restassured.mapper.DataToDeserialize
[WARNING]   io.restassured.mapper.ObjectDeserializationContext
[WARNING]   io.restassured.mapper.TypeRef
[WARNING]   io.restassured.mapper.factory.ObjectMapperFactory
[WARNING]   io.restassured.mapper.resolver.ObjectMapperResolver
[WARNING] Found duplicate classes/resources in test classpath.
```

markkolich commented:
@johanhaleby rebased with the latest `upstream/master`, including the changes in https://github.com/rest-assured/rest-assured/pull/1137 from @azzazzel.

markkolich commented on rest-assured/pom.xml:
```diff
@@ -49,8 +49,9 @@
                         	!io.restassured.path.xml.*,
-                        	!io.restassured.path.json.*,
-                        	!io.restassured.common.*,
+                            !io.restassured.path.json.*,
```
@johanhaleby @azzazzel @ponziani hi, this is required to get the maven duplicate finder happy again, but the OSGi tests in this project still fail:

https://travis-ci.org/rest-assured/rest-assured/jobs/494305823

Is this the right thing to do? How do we get the OSGi tests fixed up?

azzazzel replied:
I'll look into it.

azzazzel replied:
I just sent a PR to you.
See https://github.com/markkolich/rest-assured/pull/1

markkolich commented:
@johanhaleby I think this is ready to go! I cleaned this up and squashed @azzazzel latest changes to get this unblocked. There's a few dependency cleanups I made as well given the failures from the maven duplicate finder plugin.

johanhaleby commented:
Thanks a lot for all your help and efforts. Really appreciated.
